From: Marcelo Tosatti <mtosatti@redhat.com>
Date: Wed, 27 Jan 2010 21:13:15 -0500
Subject: [kvm] fix spurious interrupt with irqfd
Message-id: <20100127211711.891642336@amt.cnet>
Patchwork-id: 22964
O-Subject: [RHEL 6 KVM 9/9] KVM: fix spurious interrupt with irqfd
Bugzilla: 559343
RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>
RH-Acked-by: Gleb Natapov <gleb@redhat.com>

commit df9258d4671281d5c075ddaa3c4c6c820fd9f66a
Author: Michael S. Tsirkin <mst@redhat.com>
Date:   Wed Jan 13 19:12:30 2010 +0200

KVM: fix spurious interrupt with irqfd

kvm didn't clear irqfd counter on deassign, as a result we could get a
spurious interrupt when irqfd is assigned back. this leads to poor
performance and, in theory, guest crash.

Signed-off-by: Michael S. Tsirkin <mst@redhat.com>
Signed-off-by: Avi Kivity <avi@redhat.com>
Bugzilla: 559343

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/virt/kvm/eventfd.c b/virt/kvm/eventfd.c
index 942702a..486c604 100644
--- a/virt/kvm/eventfd.c
+++ b/virt/kvm/eventfd.c
@@ -72,12 +72,13 @@ static void
 irqfd_shutdown(struct work_struct *work)
 {
 	struct _irqfd *irqfd = container_of(work, struct _irqfd, shutdown);
+	u64 cnt;
 
 	/*
 	 * Synchronize with the wait-queue and unhook ourselves to prevent
 	 * further events.
 	 */
-	remove_wait_queue(irqfd->wqh, &irqfd->wait);
+	eventfd_ctx_remove_wait_queue(irqfd->eventfd, &irqfd->wait, &cnt);
 
 	/*
 	 * We know no new events will be scheduled at this point, so block
